### DISCUSSION TITLE
x86-s

### DISCUSSION THREAD

**Discussion by bartoldeman** [dosemu2/dosemu2#2010] [General]
This is "interesting", and probably something inevitable, but not good for dosemu...

https://www.intel.com/content/www/us/en/developer/articles/technical/envisioning-future-simplified-architecture.html

Still in the design stages, but future Intel chips may only run 64-bit kernels and allow limited 32-bit flat user space. No modify_ldt, ioperm, or iopl, no segmentation/VM86/etc in KVM either any more. This mean both KVM and native DPMI will stop working for us, and no more direct hardware access, which is very fragile today already.

**Comment by stsp** [comment]
Hi Bart, no, this article explicitly says
VMX will remain, hence the KVM.
There are already some things that
are available only under VMX, though
I currently can't recall what was those.

**Reply by bartoldeman** [reply]
see 3.20.3 Legacy OS Virtualization in the PDF, it specifically says that VMX needs emulation for anything vm86, real mode, 16bit or segmented, etc.

**Reply by bartoldeman** [reply]
And I'd be surprised if KVM devs would provide all that emulation in kernel space, though it's not impossible, there already is some.

**Reply by stsp** [reply]
Yes, seems you are right.
That's really weird: you either have
x86 with all compatibility (at least
under VTx), or you do ARM.
Doing a reduced set on top of x86
sounds plain crazy to me.

**Reply by stsp** [reply]
> I currently can't recall what was those.

I think that was v86, which, under VTx,
IIRC works even in 64bit mode, while
on the host it doesn't.

**Reply by stsp** [reply]
I think the only real problem here is
that JIT will break. It generates the
32bit instructions currently, so another
back-end will be needed. But the same
is with arm for example.
I really think if we move to tcg, we
will not even notice this "problem".

**Reply by bartoldeman** [reply]
I think JIT won't be affected, they are just regular x86-64 instructions, there is nothing relying on segmentation there, and addressing is always 64bit (yes you see `// movb offs(%%ebx),%%al` all over the place but it means `// movb offs(%%rbx),%%al` in 64bit.

When I ported JIT to 64bit I just took advantage of the fact that the default operand size in 64b is still 32bits, whereas default addressing size is 64bit, and so a lot of the generated binary code can be identical.

**Reply by stsp** [reply]
Even better then. :)
Yes, I forgot that CS is not changed by jit,
only native DPMI does.

**Reply by stsp** [reply]
Hmm, they seem to say compatibility
mode remains. Though it might be lacking
16bit, hence no native DPMI, correct?

**Comment by stsp** [comment]
But in a bigger scale - yes, we need to
invest into JIT for portability and future-proofs.

**Comment by Torinde** [comment]
[x86s-EAS-v1-4-17-23-1.pdf](https://cdrdv2.intel.com/v1/dl/getContent/776648) section "3.20.3 Legacy OS Virtualization" has detailed list of features that "A VMM can choose to emulate legacy functionality as required" and "Cases that are un-virtualizable":
```
a. RFLAGS.IOPL != 0: When IOPL is non-zero most cases where behavior would
typically change will instead #GP, which the VMM can catch/emulate (i.e., many
cases are virtualizable). The problematic Intel64 cases are as follows.

   i. Ring 0 privileged SW sets IOPL to 3 and changes to ring 3. If ring 3 SW
runs PUSHF or SYSCALL, the value with IOPL=3 should go into the
memory or register destination. When this sequence runs in a VM, the ring
0 instruction that sets IOPL to 3 would cause #GP and trigger a VMExit. If
the VMM resumes the VM with the “wrong” IOPL, i.e., IOPL==0, the ring 3
PUSHF or SYSCALL would expose this incorrect IOPL through memory or
the register. Also, ring3 POPF will not update IF. The preferred scheme is
for VMM to emulate the guest till IOPL is changed back to 0. This case is
not expected on a modern OS.

   ii. If the guest attempts to set IOPL to a value greater than zero using a
POPF instruction in ring 0, this will be silently ignored. The IOPL value will
not be updated and the VMM will be unaware that this occurred. Some
subsequent consumers of this value will #GP (e.g., CLI/STI/IN/OUT) but
others will silently continue with different semantics (e.g., IF updating
POPF, memory written by PUSHF, flags stored by SYSCALL, etc.)

b. #UD behavior on SYSCALL/SYSEXIT when EFER.SCE is cleared.
```

**Reply by stsp** [reply]
This isn't helping much, because
both v86 and DPMI would be removed
for no 16bit segments support.
IOPL stuff is just a minor detail.

**Reply by Torinde** [reply]
Above I pasted only the "un-virtualizable" features. The other part of the document states that "A VMM can choose to emulate legacy functionality as required" where it says v86 can be emulated. Same as your comments further up. So, with modifications, it'll be eventually possible to make, correct?

**Reply by stsp** [reply]
Not sure what you mean by "eventually
possible".
We already have the CPU emulator and
the JIT, so things do work and will remain
so. KVM is providing an acceleration, and
it will stop doing that.

**Reply by Torinde** [reply]
I meant to keep the acceleration by providing some kind of "KVM shim/mod", which emulates the functionality removed by x86-S.

**Reply by stsp** [reply]
I am not sure its a viable plan.
v86 needs full emulation.
DPMI needs full emulation as long
as at least 1 16bit segment is used
(which is almost always the case).
Its just a dead horse, really.
We need to invest into JIT, at least
that exists and works already.

**Reply by stsp** [reply]
You can as well ask kvm devs to
provide the software emulation to
not break the existing KVM API.
It will work much faster if handled
in the kernel. But I am exceptionally
skeptical they are going to do that.

**Reply by Torinde** [reply]
x86-S document doesn't mention System Management Mode as being dropped, can that help? To make it easier to re-implement CSM or VGA features?

**Reply by stsp** [reply]
No, it doesn't help.
The attempts to emulate VGA with
SMM are rumored to be attempted
by Intel, but I've never seen the outcome.
Maybe some chipsets could do that,
but hardly anyone seen such.
Also x86-S has nothing to do with the
VGA drop, because its even worse than
the VGA drop. VGA drop can be overcame
with a VM for UEFI, but with x86-S such
VM will need to use cpu emulation.
So with x86-S there is no other choice
but to use qemu or dosemu2 or dosbox-X
for running DOS, no matter what.
The question is only where to put them.
Having them as an UEFI apps may save
some boot time, but will require a horrible
amount of the porting work.
